### DISCUSSION TITLE
[Proposal] FFmpeg call handing refactoring and FFmpeg remote integration

### DISCUSSION THREAD

**Discussion by JPVenson** [jellyfin/jellyfin-meta#36] [Ideas]
# [The proposal is now out of draft status and can be discussed]

Hey,

I would like to propose 2 things that are tight together. As Tim (@TimGels) and I have extensively on discord discussed, we are aiming to provide a more tightly integrated solution to remotely transcode and run ffmpeg. We are aware that there are existing solutions (rffmpeg) but they are not as user friendly as it could be and due to the generalized way are missing on some comfort features then a directly implemented solution.

My proposal is divided into two parts:
1. Refactor all ffmpeg calls into a structured and object oriented way
2. Build a solution that allows the user to manage and jellyfin to invoke ffmpeg on a remote machines

The 2nd point cannot be realized in the existing environment as discussed later.

# 1. ffmpeg refactoring
 As of today there are XXX instances where the ffmpeg executable is either called directly without capsulation, or through statically typed helper methods. 
- `Emby.Server.Implementations.LiveTv.EmbyTV.EncodedRecorder` 
Calls the ffmpeg executable directly with some very specific arguments to transcode some media source
- `MediaBrowser.Controller.MediaEncoding.IMediaEncoder` implementation `MediaBrowser.MediaEncoding.Encoder.MediaEncoder`
Is an encapsulation for the ffmpeg process, does however expose filesystem references to ffmpeg that are used at the `TranscodingJobHelper` as well as the `AttachmentExtractor` and `SubtitleEncoder`
- `Jellyfin.Api.Helpers.TranscodingJobHelper`
Is referenced statically from several api controllers and helpers and therefor cannot be replaced at runtime.

I propose a encapsulation of all direct or indirect calls to ffmpeg via a `JellyfinFfmpegService`. This service provides a structured way constructing ffmpeg commands similar to [FFMpegCore](https://github.com/rosenbjerg/FFMpegCore#ffmpeg). However this implementation would not directly allow specification of primary output data but only provide a new `JellyfinFfmpegStream` or similar to access the result. With this abstraction it will be possible to later virtualize local and remote ffmpeg calls. Also by providing this low level of I/O abstraction, most ffmpeg specific code can be retained with minimal adaptation. 

There are 2 main cases how jellyfin handles the ffmpeg output:
- a. Save to Storage (examples: Subtitle Encoding, Chapter Generation)
- b. Provide via Api (Hls Streaming)

For a. the `JellyfinFfmpegStream` will provide a method `RealiseToStorageAsync(MediaSourceInfo | FileInfo)` to obtain the complete stream on the requested storage location. For b. the stream would allow partial, byte exact transmission either on demand or preemptively by accessing the `System.IO.Stream` methods the `JellyfinFfmpegStream` implements. This might need some rework on the HLS controller to be optimal as the current FileStream access buffers might be ineffective in this networked scenario. 

The whole goal of this refactoring is to enable an in-process handling of ffmpeg requests to be serialized and executed remotely. Currently this can only be done by replacing the FFMpeg executable which makes this non-trivial for non-technical users as well as missing out on some benefits a tight integration would bring, such as the ability to handle all transcoding in memory on the jellyfin server, as in an on-demand scenario only actively requested data will be requested externally, which in turn would reduce the disk interactions drastically as well as gains in responsiveness as the JF hosting service would be freed from any processing and be focused to only serve the main application.

# 2. Build a Remote FFMpeg solution
While the road to refactoring the ffmpeg access is relatively straight in terms of scope, the proposal for a `JellyfinRemoteFFmpegServer` (JRFS) is a bit more unclear to me. 
The idea would be to either integrate a system into jellyfin-server directly that allows to manage multiple services or with the help of above refactoring build the system on a plugin basis. 

The general inner workings would stay the same and as follows: When a ffmpeg job is requested, the request would be serialized and transmitted to an external server. This server would consist of an aspcore runtime hosting a WebApi 2.0 and would have a jellyfin FFmpeg executable located. The Serialized request would then be processed and provided back to the requesting jellyfin server. 

Data transfer of the source media files will be handled in the 1st iteration of JRFS by requiring direct access to the storage medium. This can be achieved by ether mounting a remote location to the corresponding JRFS or providing physical access to the files. In the 2nd iteration the JRFS will also support other means of obtaining the media files such as from jf-server directly.

This poses two important questions. Should the JRFS _push_ the requested output back the the jf-server or should the jf-server _pull_ such data on-demand. Both settings have pro and cons.

### Push
Pro:
- All requested data would be available immediately when requested by a client causing only negligible delay for transcoding in some cases

Con:
- The Jf-server network would be more congested while streaming as data comes in from the JRFS and goes out to the clients
- Most advantages due to not storing transcoding files would be negated

### Pull
Pro:
- No I/O on the storage of the jf-server
- No unnecessary I/O on the network

Con:
- Expected additional Delay in requesting transcoded chunks of data as a request to the JRFS must be made and be processed
- When a temporary network problem arises the effects will be immediately noticeable leading to lag.

In both cases the JRFS aspcore server will be the interface to access any data that is beeing processed.

The 1st iteration should implement any one or both of the options but for a 2nd i am thinking of a peer-to-peer enablement. This would be exclusively for Transcoding jobs that are requested by any client. The idea is to start a remote process and then instead transferring the data intermediately to the jf-server, to point any requesting client to the corresponding JRFS server. I plan to do this by modifying the corresponding api controller and return a 302 to the correct service. However this has some considerable implications, as the required networking setup for the non-technical Enduser is significant, as all JRFS servers must be reachable by the requesting client. This would also facilitate a great number of modifications in the web-api, more or less requiring that the JRFS is a 1st party supported feature.

The JRFS would be firstly provided as a docker container to be installed and configured via environment variables and/or a configuration file. It would register itself at the configured jellyfin-server instance and from then on be available as a ffmpeg processor. For the 1st iteration it is only planned to allow one JRFS to be registered, later iterations might introduce multiple instances with load balancing.

# Questions Open
- Would a refactoring as described for the ffmpeg access have a chance to be accepted 
- Should such remote ffmpeg functionality be part of jellyfin or better be a 3rd party integration
- Should the integration be based on Push/Pull/Both/Other 
- Is a Peer-To-Peer solution viable or not or only as a "pro-user" feature

**Comment by sevenrats** [comment]
this is a great idea!

**Comment by gnattu** [comment]
I hate to be that guy but:

> Would a refactoring as described for the ffmpeg access have a chance to be accepted

No. Reasons will be explained below. The proposed design is not realistic and not that useful to archive the design goal. An alternative design in my mind will be presented below as well.

> Should such remote ffmpeg functionality be part of jellyfin or better be a 3rd party integration

We can have both, if we can make the api stable enough. And if we abstract the server higher level enough, People with enough interest can even write alternative transcoding services that is not even ffmpeg based.

> Should the integration be based on Push/Pull/Both/Other

Other. Reasons below.

> Is a Peer-To-Peer solution viable or not or only as a "pro-user" feature

Not viable at all, unless you want to implement something like userspace NFS within jellyfin that looks too out of scope to me.


--------

I understand the use case for a remote transcoding service but what we need is not an ffmpeg wrapper and a single `JellyfinFfmpegStream` is also not easy to deal with because hls streams are not a single file, it is a series of files, and combining and separating those files within a single stream is over-engineering to me.

FFmpeg, at least in the context of jellyfin, isn’t an executable on its own. It’s designed to run in a tightly packed environment that includes the OS kernel, drivers from both kernel space and user space, the hardware, and its supporting runtime. Consequently, the transcoder’s behavior varies significantly across different environments. For instance, the command generated on different Linux kernel versions will differ, so as the command generated on macOS, Windows, or Linux. Additionally, the command generated with different runtime versions will also be different. This is what rffmpeg currently failed to handle and is already causing issues for users attempting to use that approach because the transcode service generated wrong command as it wrongly interpreted the server is running on an unaffected kernel, or the command is outright wrong because the remote machine is on Windows but the server runs Linux. If we want to handle this use case seriouly we don't abstract ffmpeg call because that did not work well and will not work well.

What in my mind is to abstract the whole transcoding service (basically the whole `MediaEncoder` and its `EncodingHelpers`) so that it can on its own and not necessarily part of the server. We define our tasks better, and use standard API calls for transcoding servers to execute that "task", and use pre-defined api endpoints to get the result file back from the transcoding server. This would work for both media transcoding and static file processing. We need to carefully design said APIs and the callbacks to handle the interaction between server and transcoders. This would also allow the server to have multiple transcoders each with different hardware and software combinations, and having hardware transcoding settings on their own if we can design a work distribution system, or just use state of the art implementations like rabbitMQ or Kafka, if that is not an overkill.

The media source files should not be streamed from the main server because ffmpeg works best with filesystems inputs directly. Any kind of remote sources handled by ffmpeg will introduce a significant delay that we want to avoid. So we need to require said transcode servers to have its own mount point of the media files so that it can find the files on its own mount point. Each said transcoding server should have its own transcoding cache and it serves files from its own cache via predefined api. The api could server a file (for transcoded files or extracted attachements), or provide a json (for file info probe). 

And for such design, obviously, we don't need to do ffmpeg calling abstraction at all because we don't even care if the transcoding is done by ffmpeg. It can literally be anything, as long as the service wrapping it communicates with the server in our defined way. I think this works better and is more flexible.

**Comment by nyanmisaka** [comment]
This wrapper [cannot handle `-filter_complex`](https://github.com/rosenbjerg/FFMpegCore/issues/422), causing `WithCustomArgument()` to have to be used. There are many similar use cases such as encoder options that do not use the global AVoptions of ffmpeg. As a result, we still have to use a large number of custom arguments, which makes its security not much better than what we have now.

Ideally, we should abstract and define the codecs and filters we rely on, so that MediaEncoder callers can only use encapsulated component functions and enumeration values ​​and filter out suspicious strings.

I recently resumed the task of adding hardware information probing to ffprobe, which I believe will eventually lead to a refactoring of the transcoder, especially in terms of avoiding abusive use of string concatenation.
